warehouse/internal/repo/staging.go

```diff
@@ -0,0 +1,288 @@
+		SourceJobID:           stagingFile.SourceJobID,
+		SourceJobRunID:        stagingFile.SourceJobRunID,
+		TimeWindowYear:        stagingFile.TimeWindow.Year(),
+		TimeWindowMonth:       int(stagingFile.TimeWindow.Month()),
+		TimeWindowDay:         stagingFile.TimeWindow.Day(),
+		TimeWindowHour:        stagingFile.TimeWindow.Hour(),
+		DestinationRevisionID: stagingFile.DestinationRevisionID,
+	}
+}
+
+func (m *metadataSchema) SetStagingFile(stagingFile *model.StagingFile) {
+	stagingFile.UseRudderStorage = m.UseRudderStorage
+	stagingFile.SourceBatchID = m.SourceBatchID
+	stagingFile.SourceTaskID = m.SourceTaskID
+	stagingFile.SourceTaskRunID = m.SourceTaskRunID
+	stagingFile.SourceJobID = m.SourceJobID
+	stagingFile.SourceJobRunID = m.SourceJobRunID
+	stagingFile.TimeWindow = time.Date(m.TimeWindowYear, time.Month(m.TimeWindowMonth), m.TimeWindowDay, m.TimeWindowHour, 0, 0, 0, time.UTC)
+	stagingFile.DestinationRevisionID = m.DestinationRevisionID
+}
+
+func (repo *StagingFiles) init() {
+	repo.once.Do(func() {
+		if repo.Now == nil {
+			repo.Now = timeutil.Now
+		}
+	})
+}
+
+// Insert inserts a staging file into the staging files table. It returns the ID of the inserted staging file.
+//
+// NOTE: The following fields are ignored and set by the database:
+// - ID
+// - Error
+// - CreatedAt
+// - UpdatedAt
+func (repo *StagingFiles) Insert(ctx context.Context, stagingFile *model.StagingFile) (int64, error) {
+	repo.init()
+
+	var (
+		id                        int64
+		firstEventAt, lastEventAt interface{}
+	)
+
+	firstEventAt = stagingFile.FirstEventAt.UTC()
+	if stagingFile.FirstEventAt.IsZero() {
+		firstEventAt = nil
+	}
+
+	lastEventAt = stagingFile.LastEventAt.UTC()
+	if stagingFile.LastEventAt.IsZero() {
+		lastEventAt = nil
+	}
+
+	m := metadataFromStagingFile(stagingFile)
+	rawMetadata, err := json.Marshal(&m)
+	if err != nil {
+		return id, fmt.Errorf("marshaling metadata: %w", err)
+	}
+	now := repo.Now()
+
+	schemaPayload, err := json.Marshal(stagingFile.Schema)
+	if err != nil {
+		return id, fmt.Errorf("marshaling schema: %w", err)
+	}
+
+	err = repo.DB.QueryRowContext(ctx,
+		`INSERT INTO `+tableName+` (
+			location,
+			schema,
+			workspace_id,
+			source_id,
+			destination_id,
+			status,
+			total_events,
+			first_event_at,
+			last_event_at,
+			created_at,
+			updated_at,
+			metadata
+		)
+		VALUES
+		 ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12) 
+		RETURNING id`,
+
+		stagingFile.Location,
+		schemaPayload,
+		stagingFile.WorkspaceID,
+		stagingFile.SourceID,
+		stagingFile.DestinationID,
+		stagingFile.Status,
+		stagingFile.TotalEvents,
+		firstEventAt,
+		lastEventAt,
+		now.UTC(),
+		now.UTC(),
+		rawMetadata,
+	).Scan(&id)
+	if err != nil {
+		return id, fmt.Errorf("inserting staging file: %w", err)
+	}
+
+	return id, nil
+}
+
+// praseRow is a helper for mapping a row of tableColumns to a model.StagingFile.
+func (*StagingFiles) parseRows(rows *sql.Rows) ([]model.StagingFile, error) {
```

Comment:
We are not calling rows.close() anywhere?

Reply:
```suggestion
func (*StagingFiles) parseRows(rows *sql.Rows) ([]model.StagingFile, error) {
	var (
		stagingFiles []model.StagingFile
		err          error
	)

	defer func() {
		closeErr := rows.Close()
		if err == nil {
			if closeErr != nil {
				fmt.Println("closing rows: ", closeErr)
			}
		}
		err = closeErr
	}()
```

Reply:
```suggestion
func (*StagingFiles) parseRows(rows *sql.Rows) ([]model.StagingFile, error) {
	var (
		stagingFiles []model.StagingFile
		err          error
	)

	defer func() {
		closeErr := rows.Close()
		if err == nil {
			if closeErr != nil {
				fmt.Println("closing rows: ", closeErr)
			}
		}
		err = closeErr
	}()
```

Reply:
I have seen that in most of the places in our code base, we are not handling close errors. Maybe it's a good time to handle it properly.

Reply:
https://github.com/teivah/100-go-mistakes/blob/4fdb88c13ed09f626863d40e9532b39628f79889/07-error-management/54-defer-errors/main.go#L32

Reply:
@achettyiitr, you are right and I was missing error handling as well.

After some thought, I think this is a simpler equivalent would be:

```go
	defer func() { _ = rows.Close() }()

	for rows.Next() {
		var (
			stagingFile model.StagingFile
			metadataRaw []byte

			firstEventAt, lastEventAt sql.NullTime

			errorRaw sql.NullString
		)
		err := rows.Scan(
			&stagingFile.ID,
			&stagingFile.Location,
			&stagingFile.SourceID,
			&stagingFile.DestinationID,
			&stagingFile.Schema,
			&errorRaw,
			&stagingFile.Status,
			&firstEventAt,
			&lastEventAt,
			&stagingFile.TotalEvents,
			&stagingFile.CreatedAt,
			&stagingFile.UpdatedAt,
			&metadataRaw,
			&stagingFile.WorkspaceID,
		)
		if err != nil {
			return nil, fmt.Errorf("scanning row: %w", err)
		}

		stagingFile.CreatedAt = stagingFile.CreatedAt.UTC()
		stagingFile.UpdatedAt = stagingFile.UpdatedAt.UTC()

		if firstEventAt.Valid {
			stagingFile.FirstEventAt = firstEventAt.Time.UTC()
		}
		if lastEventAt.Valid {
			stagingFile.LastEventAt = lastEventAt.Time.UTC()
		}

		if errorRaw.Valid {
			stagingFile.Error = errors.New(errorRaw.String)
		}

		var m metadataSchema
		err = json.Unmarshal(metadataRaw, &m)
		if err != nil {
			return nil, fmt.Errorf("unmarshal metadata: %w", err)
		}

		m.SetStagingFile(&stagingFile)
		stagingFiles = append(stagingFiles, stagingFile)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating rows: %w", err)
	}
```

Reason, `.Close` will be effective only if we interrupt the iteration. 

If `.Next()` returns false it will close everything automatically; in this case, we only need to check for an error.

The only case we need to close is when scan or unmarshal fails. In that case, we would already have an error. In the example of the book, if there is already an error is not getting overwritten by a close error. I think this is the correct approach; close error is less significant than the others and combining the two errors might lead to confusion.

Reply:
Thanks, @lvrach for sharing thoughts on this.